File: ui/src/schemas/exports/normal-schema.ts
/**
 * Normal Map Export Schemas
 *
 * Zod schemas for validating normal map export data.
 * Matches the exact structure returned by BackendDepthService.generateNormal().
 */

import { z } from "zod";

// ============================================================================
// Primitives
// ============================================================================

const positiveInt = z.number().int().positive();
const positiveFinite = z.number().finite().positive();

// ============================================================================
// Normal Generation Options Schema
// ============================================================================

export const NormalGenerationMethodSchema = z.enum(["algebraic", "normalcrafter"]);

export type NormalGenerationMethod = z.infer<typeof NormalGenerationMethodSchema>;

export const NormalDepthModelSchema = z.enum([
  "DA3-LARGE-1.1",
  "DA3-GIANT-1.1",
  "DA3NESTED-GIANT-LARGE-1.1",
]);

export type NormalDepthModel = z.infer<typeof NormalDepthModelSchema>;

export const NormalGenerationOptionsSchema = z.object({
  method: NormalGenerationMethodSchema.optional(),
  depthModel: NormalDepthModelSchema.optional(),
});

export type NormalGenerationOptions = z.infer<typeof NormalGenerationOptionsSchema>;

// ============================================================================
// Normal Generation Result Schema
// ============================================================================

/**
 * Normal map generation result.
 * Matches the exact structure returned by BackendDepthService.generateNormal().
 */
export const NormalGenerationMetadataSchema = z.object({
  method: z.string().min(1),
  width: positiveInt,
  height: positiveInt,
});

export type NormalGenerationMetadata = z.infer<typeof NormalGenerationMetadataSchema>;

export const NormalGenerationResultSchema = z.object({
  status: z.enum(["success", "error"]),
  normal: z.string(), // base64 encoded PNG (RGB normal map)
  depth: z.string().optional(), // base64 depth map used (if generated)
  fallback: z.boolean().optional(),
  message: z.string().optional(),
  metadata: NormalGenerationMetadataSchema.optional(),
});

export type NormalGenerationResult = z.infer<typeof NormalGenerationResultSchema>;

// ============================================================================
// Validation Helpers
// ============================================================================

export function validateNormalGenerationResult(
  data: unknown,
): NormalGenerationResult {
  return NormalGenerationResultSchema.parse(data);
}

export function safeValidateNormalGenerationResult(data: unknown) {
  return NormalGenerationResultSchema.safeParse(data);
}

File: lattice-core/haskell/Lattice/Schemas/Exports/NormalSchema.hs
{-# LANGUAGE StrictData #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE DerivingStrategies #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Module      : Lattice.Schemas.Exports.NormalSchema
Description : Normal map export format enums and types
Copyright   : (c) Lattice, 2026
License     : MIT

Normal map export format enums and types.

Source: ui/src/schemas/exports/normal-schema.ts
-}

module Lattice.Schemas.Exports.NormalSchema
  ( -- * Normal Generation Method
    NormalGenerationMethod(..)
  , normalGenerationMethodFromText
  , normalGenerationMethodToText
    -- * Normal Depth Model
  , NormalDepthModel(..)
  , normalDepthModelFromText
  , normalDepthModelToText
    -- * Generation Status
  , GenerationStatus(..)
  , generationStatusFromText
  , generationStatusToText
    -- * Constants
  , maxDimension
    -- * Structures
  , NormalGenerationOptions(..)
  , NormalGenerationMetadata(..)
  , NormalGenerationResult(..)
    -- * Validation
  , isValidMetadata
  ) where

import GHC.Generics (Generic)
import Data.Text (Text)

-- ────────────────────────────────────────────────────────────────────────────
-- Normal Generation Method
-- ────────────────────────────────────────────────────────────────────────────

-- | Normal generation method options
data NormalGenerationMethod
  = MethodAlgebraic
  | MethodNormalcrafter
  deriving stock (Eq, Show, Generic, Enum, Bounded)

normalGenerationMethodFromText :: Text -> Maybe NormalGenerationMethod
normalGenerationMethodFromText "algebraic" = Just MethodAlgebraic
normalGenerationMethodFromText "normalcrafter" = Just MethodNormalcrafter
normalGenerationMethodFromText _ = Nothing

normalGenerationMethodToText :: NormalGenerationMethod -> Text
normalGenerationMethodToText MethodAlgebraic = "algebraic"
normalGenerationMethodToText MethodNormalcrafter = "normalcrafter"

-- ────────────────────────────────────────────────────────────────────────────
-- Normal Depth Model
-- ────────────────────────────────────────────────────────────────────────────

-- | Normal depth model options
data NormalDepthModel
  = DA3_LARGE_1_1
  | DA3_GIANT_1_1
  | DA3NESTED_GIANT_LARGE_1_1
  deriving stock (Eq, Show, Generic, Enum, Bounded)

normalDepthModelFromText :: Text -> Maybe NormalDepthModel
normalDepthModelFromText "DA3-LARGE-1.1" = Just DA3_LARGE_1_1
normalDepthModelFromText "DA3-GIANT-1.1" = Just DA3_GIANT_1_1
normalDepthModelFromText "DA3NESTED-GIANT-LARGE-1.1" = Just DA3NESTED_GIANT_LARGE_1_1
normalDepthModelFromText _ = Nothing

normalDepthModelToText :: NormalDepthModel -> Text
normalDepthModelToText DA3_LARGE_1_1 = "DA3-LARGE-1.1"
normalDepthModelToText DA3_GIANT_1_1 = "DA3-GIANT-1.1"
normalDepthModelToText DA3NESTED_GIANT_LARGE_1_1 = "DA3NESTED-GIANT-LARGE-1.1"

-- ────────────────────────────────────────────────────────────────────────────
-- Generation Status
-- ────────────────────────────────────────────────────────────────────────────

-- | Generation status options
data GenerationStatus
  = StatusSuccess
  | StatusError
  deriving stock (Eq, Show, Generic, Enum, Bounded)

generationStatusFromText :: Text -> Maybe GenerationStatus
generationStatusFromText "success" = Just StatusSuccess
generationStatusFromText "error" = Just StatusError
generationStatusFromText _ = Nothing

generationStatusToText :: GenerationStatus -> Text
generationStatusToText StatusSuccess = "success"
generationStatusToText StatusError = "error"

-- ────────────────────────────────────────────────────────────────────────────
-- Constants
-- ────────────────────────────────────────────────────────────────────────────

maxDimension :: Int
maxDimension = 16384

-- ────────────────────────────────────────────────────────────────────────────
-- Structures
-- ────────────────────────────────────────────────────────────────────────────

-- | Normal generation options
data NormalGenerationOptions = NormalGenerationOptions
  { ngoMethod :: !(Maybe NormalGenerationMethod)
  , ngoDepthModel :: !(Maybe NormalDepthModel)
  }
  deriving stock (Eq, Show, Generic)

-- | Normal generation metadata
data NormalGenerationMetadata = NormalGenerationMetadata
  { ngmMethod :: !Text
  , ngmWidth :: !Int
  , ngmHeight :: !Int
  }
  deriving stock (Eq, Show, Generic)

-- | Normal generation result
data NormalGenerationResult = NormalGenerationResult
  { ngrStatus :: !GenerationStatus
  , ngrNormal :: !Text        -- ^ Base64 encoded PNG (RGB normal map)
  , ngrDepth :: !(Maybe Text) -- ^ Base64 depth map used (if generated)
  , ngrFallback :: !(Maybe Bool)
  , ngrMessage :: !(Maybe Text)
  , ngrMetadata :: !(Maybe NormalGenerationMetadata)
  }
  deriving stock (Eq, Show, Generic)

-- ────────────────────────────────────────────────────────────────────────────
-- Validation
-- ────────────────────────────────────────────────────────────────────────────

-- | Check if metadata is valid
isValidMetadata :: NormalGenerationMetadata -> Bool
isValidMetadata m =
  ngmWidth m > 0 && ngmWidth m <= maxDimension &&
  ngmHeight m > 0 && ngmHeight m <= maxDimension
